AMM/contracts/AMM/tests/utils/mod.rs

```diff
@@ -0,0 +1,152 @@
+use fuels::{contract::contract::CallResponse, prelude::*};
+use rand::prelude::Rng;
+use std::str::FromStr;
+
+abigen!(AMM, "out/debug/amm-abi.json");
+abigen!(Exchange, "../exchange/out/debug/exchange-abi.json");
+
+pub mod exchange_abi_calls {
+    use super::*;
+
+    pub async fn constructor(
+        contract: &Exchange,
+        pair: (ContractId, ContractId),
+    ) -> CallResponse<()> {
+        let receipt = contract.methods().constructor(pair).call().await;
+        dbg!(&receipt);
+        receipt.unwrap()
+    }
+}
+
+pub mod amm_abi_calls {
+    use super::*;
+
+    pub async fn initialize(contract: &AMM, exchange_contract_id: ContractId) -> CallResponse<()> {
+        contract
+            .methods()
+            .initialize(exchange_contract_id)
+            .call()
+            .await
+            .unwrap()
+    }
+
+    pub async fn add_pool(
+        contract: &AMM,
+        asset_pair: (ContractId, ContractId),
+        pool: ContractId,
+    ) -> CallResponse<()> {
+        contract
+            .methods()
+            .add_pool(asset_pair, pool)
+            .set_contracts(&[pool.into()])
+            .call()
+            .await
+            .unwrap()
+    }
+
+    pub async fn pool(contract: &AMM, asset_pair: (ContractId, ContractId)) -> Option<ContractId> {
+        contract
+            .methods()
+            .pool(asset_pair)
+            .call()
+            .await
+            .unwrap()
+            .value
+    }
+}
+
+pub mod paths {
+    pub const AMM_CONTRACT_BINARY_PATH: &str = "out/debug/amm.bin";
+    pub const AMM_CONTRACT_STORAGE_PATH: &str = "out/debug/amm-storage_slots.json";
+    pub const INVALID_EXCHANGE_CONTRACT_BINARY_PATH: &str =
```

Comment:
It is used for an implementation of the exchange contract ABI that is not registered to the AMM, so that even if the pairs match, the exchange contract cannot be used in the AMM.

Reply:
Could we use a valid contract but salt it so that a different root is returned or would that not work? If this works then that file is unnecessary.

Reply:
I believe not, given that `test functions::add_pool::success::adds_more_than_once` works.

Reply:
Note: I missed the "valid / invalid" naming here. Changed it to use "malicious" instead.